Xamarin.Forms.Core/Shell/Shell.cs

```diff
@@ -1653,6 +1653,47 @@ bool ProposeNavigation(ShellNavigationSource source, ShellNavigationState propos
 			return !navArgs.Cancelled && navArgs.DeferralCount == 0;
 		}
 
+		ShellNavigationSource CalculateNavigationSource(ShellNavigationState current, NavigationRequest request)
+		{
+			if (request.StackRequest == NavigationRequest.WhatToDoWithTheStack.PushToIt)
+				return ShellNavigationSource.Push;
+
+			if (current == null)
+				return ShellNavigationSource.Unknown;
+
+			var targetUri = ShellUriHandler.ConvertToStandardFormat(this, request.Request.FullUri);
+			var currentUri = ShellUriHandler.ConvertToStandardFormat(this, current.FullLocation);
+			
+			var targetPaths = ShellUriHandler.RetrievePaths(targetUri.PathAndQuery);
+			var currentPaths = ShellUriHandler.RetrievePaths(currentUri.PathAndQuery);
+
+			if (targetPaths.Length < 4 || currentPaths.Length < 4)
+				return ShellNavigationSource.Unknown;
+
+			if (targetPaths[1] != currentPaths[1])
+				return ShellNavigationSource.ShellItemChanged;
+			if (targetPaths[2] != currentPaths[2])
+				return ShellNavigationSource.ShellSectionChanged;
+			if (targetPaths[3] != currentPaths[3])
+				return ShellNavigationSource.ShellContentChanged;
+
+			if (targetPaths.Length == currentPaths.Length)
+				return ShellNavigationSource.Unknown;
+
+			if(targetPaths.Length < currentPaths.Length)
+			{
+				if (targetPaths[targetPathsLength - 1] == currentPaths[currentPathsLength - 1])
+					return ShellNavigationSource.Remove;
+
+				return ShellNavigationSource.Pop;
+			}
+
+			if (targetPaths[targetPathsLength - 1] == currentPaths[currentPathsLength - 1])
```

Comment:
On my first suggestion I added two locals variables
```
var targetPathsLength = targetPaths.Length;
var currentPathsLength = currentPaths.Length;
```
so without them, the code will fail.

```suggestion
			if (targetPaths[targetPaths.Length - 1] == currentPaths[currentPaths.Length - 1])
```